src/com/ms/silverking/cloud/argus/Terminator.java

```diff
@@ -1,96 +1,121 @@
+    killCommands.put(name, cmd);
+  }
+
+  public Terminator(Mode mode, String loggerFileName, KillType killType, long minKillIntervalMillis) {
+    this.mode = mode;
+    runtime = Runtime.getRuntime();
+    logMessageHandler = new TerminatorAsyncLogger(loggerFileName);
+    this.killType = killType;
+    this.minKillIntervalMillis = minKillIntervalMillis;
+    lastKillMillis = new HashMap<>();
+  }
+
+  public Mode getMode() {
+    return mode;
+  }
+
+  private String[] resolvedKillCommand(String name, int pid) {
+    String unresolvedCommand;
+
+    unresolvedCommand = killCommands.get(name);
+    if (unresolvedCommand == null) {
+      Log.warning("No such kill command: " + name);
+      return null;
+    } else {
+      String resolvedCommand;
+
+      resolvedCommand = unresolvedCommand.replaceAll(PID_VARIABLE, Integer.toString(pid));
+      return resolvedCommand.split("\\s+");
     }
-    
-    private String[] resolvedKillCommand(String name, int pid) {
-        String    unresolvedCommand;
-        
-        unresolvedCommand = killCommands.get(name);
-        if (unresolvedCommand == null) {
-            Log.warning("No such kill command: "+ name);
-            return null;
-        } else {
-            String    resolvedCommand;
-            
-            resolvedCommand = unresolvedCommand.replaceAll(PID_VARIABLE, Integer.toString(pid));
-            return resolvedCommand.split("\\s+");
-        }
+  }
+
+  public void terminate(int pid, String msg) {
+    Long lastPidKillMillis;
+    long curTimeMillis;
+
+    curTimeMillis = SystemTimeUtil.skSystemTimeSource.absTimeMillis();
+    lastPidKillMillis = lastKillMillis.get(pid);
+    if (lastPidKillMillis != null) {
+      long delta;
+
+      delta = curTimeMillis - lastPidKillMillis;
+      if (delta < minKillIntervalMillis) {
+        Log.warningf("%s Last kill too soon for %d. %d < %d", mode.name(), pid, delta, minKillIntervalMillis);
+        return;
+      }
     }
-    
-    public void terminate(int pid, String msg) {
-        try {
-            String[]    cmd;
-            
-            cmd = resolvedKillCommand(killType.toString(), pid);
-            if (mode == Mode.LogOnly) {
-                Log.warning(mode.name() + " Would have run: " + StringUtil.arrayToQuotedString(cmd) + " ", pid);
-            } else {
-                runtime.exec(cmd);
-                Log.warning(mode.name() + " " + msg);
-            }
-            logMessageHandler.add(mode.name() + " " + msg);
-        } catch (IOException ioe) {
-            Log.logErrorWarning(ioe);
-        }
+    try {
+      String[] cmd;
+
+      cmd = resolvedKillCommand(killType.toString(), pid);
+      if (mode == Mode.LogOnly) {
+        Log.warning(mode.name() + " Would have run: " + StringUtil.arrayToQuotedString(cmd) + " ", pid);
+      } else {
+        lastKillMillis.put(pid, curTimeMillis);
+        runtime.exec(cmd);
+        Log.warning(mode.name() + " " + msg);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;msg&#39; (index: 1 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;pid&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, reduce the method&#39;s visibility.

Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).